File: software/cv/cv-pipeline.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <opencv2/opencv.hpp>
#include <opencv2/dnn/dnn.hpp>

using namespace cv;
using namespace dnn;

using namespace std;
//using namespace tensorflow;

int main() 
{
    
// Specify the paths for the 2 files
string protoFile = "../models/pose/mpi/pose_deploy_linevec.prototxt";
string weightsFile = "../models/pose/mpi/pose_iter_160000.caffemodel";

// Read the network into Memory
//Net net = readNetFromTensorflow("../models/movenet_multipose_lightning_1/saved_model.pb");
//cout << "Read";

  string videoArgs = "nvarguscamerasrc sensor-id=0 ! video/x-raw(memory:NVMM), width=1920, height=1080, format=(string)NV12, framerate=(fraction)20/1 ! nvvidconv flip-method=0 ! video/x-raw, width=640, height=480, format=(string)BGRx ! videoconvert ! video/x-raw, format=(string)BGR ! appsink";

  //Start capture
  VideoCapture cam(videoArgs, cv::CAP_GSTREAMER);

  if(!cam.isOpened())
    {
       printf("Camera is not opened.\n");
       return -1;
    }

    while(1)
    {

	//Executed for every frame


        Mat frame;
        cam >> frame;
        
        //cv::Size s = frame.size();
	//int frameHeight = s.height;
	//int frameWidth = s.width;
        
        // Specify the input image dimensions
	//int inWidth = 256;
	//int inHeight = 144;

// Prepare the frame to be fed to the network
	//Mat inpBlob = blobFromImage(frame, 1.0 / 255, Size(inWidth, inHeight), Scalar(0, 0, 0), false, false);

// Set the prepared object as the input blob of the network
	//net.setInput(inpBlob);
	
	//cout << "Here";
	
	//Mat output = net.forward();
	
	//cout << "There";
	
	//int H = output.size[2];
	//int W = output.size[3];
	
	//int nPairs = 18;
	///
	///Mat trimmed = output[0][0]
	

// find the position of the body parts
	//vector<Point> points(17);
	//for (int i=0; i < 17; i++) {
	//for (int n=0; n < nPairs; n++)
	//{
// Probability map of corresponding body's part.
		//Mat probMap(H, W, CV_32F, output.ptr(0,n));

		//Point2f p(-1,-1);
		//Point maxLoc;
		//double prob;
		//minMaxLoc(probMap, 0, &prob, 0, &maxLoc);
		//if (prob > .6)
		//{
		//p = maxLoc;
		//p.x *= (float)frameWidth / W ;
		//p.y *= (float)frameHeight / H ;
		
		

		//circle(frame, cv::Point((int)p.x, (int)p.y), 8, Scalar(0,255,255), -1);
		//cv::putText(frame, cv::format("%d", n), cv::Point((int)p.x, (int)p.y), cv::FONT_HERSHEY_COMPLEX, 1, cv::Scalar(0, 0, 255), 2);
		//}
		//points[n] = p;
	//}
	//}
	
//	for (int n = 0; n < nPairs; n++)
//	{
// lookup 2 connected body/hand parts
//		Point2f partA = points[POSE_PAIRS[n][0]];
//		Point2f partB = points[POSE_PAIRS[n][1]];

//	if (partA.x<=0 || partA.y<=0 || partB.x<=0 || partB.y<=0)
//continue;

//line(frame, partA, partB, Scalar(0,255,255), 8);
//circle(frame, partA, 8, Scalar(0,0,255), -1);
//circle(frame, partB, 8, Scalar(0,0,255), -1);
//}
	
	
        imshow("original",frame);
        if((char)waitKey(30) == 27)
            break;
    }
    
    return 0;
}

File: software/cv/cv-pipeline.py
# MIT License
# Copyright (c) 2019-2022 JetsonHacks

# Using a CSI camera (such as the Raspberry Pi Version 2) connected to a
# NVIDIA Jetson Nano Developer Kit using OpenCV
# Drivers for the camera and OpenCV are included in the base image

import cv2
import mediapipe as mp
import time
#import tensorflow as tf
#import numpy as np


#mpPose = mp.solutions.pose
#pose = mpPose.Pose()
#mpDraw = mp.solutions.drawing_utils

""" 
gstreamer_pipeline returns a GStreamer pipeline for capturing from the CSI camera
Flip the image by setting the flip_method (most common values: 0 and 2)
display_width and display_height determine the size of each camera pane in the window on the screen
Default 1920x1080 displayd in a 1/4 size window
"""

def gstreamer_pipeline(
    sensor_id=0,
    capture_width=1920,
    capture_height=1080,
    display_width=960,
    display_height=540,
    framerate=30,
    flip_method=0,
):
    return (
        "nvarguscamerasrc sensor-id=%d !"
        "video/x-raw(memory:NVMM), width=(int)%d, height=(int)%d, framerate=(fraction)%d/1 ! "
        "nvvidconv flip-method=%d ! "
        "video/x-raw, width=(int)%d, height=(int)%d, format=(string)BGRx ! "
        "videoconvert ! "
        "video/x-raw, format=(string)BGR ! appsink"
        % (
            sensor_id,
            capture_width,
            capture_height,
            framerate,
            flip_method,
            display_width,
            display_height,
        )
    )




def show_camera():
    window_title = "CSI Camera"

    # To flip the image, modify the flip_method parameter (0 and 2 are the most common)
    #print(gstreamer_pipeline(flip_method=0))
    video_capture = cv2.VideoCapture(gstreamer_pipeline(flip_method=0), cv2.CAP_GSTREAMER)
    while(not video_capture.isOpened()):
    	continue
    try:
      window_handle = cv2.namedWindow(window_title, cv2.WINDOW_AUTOSIZE)
      while True:
      	ret_val, frame = video_capture.read()
      	#imgRGB = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
      	#scale_percent = 20 # percent of original size
      	#width = int(img.shape[1] * scale_percent / 100)
      	#height = int(img.shape[0] * scale_percent / 100)
      	#dim = (width, height)
      	#img  = cv2.resize(img, dim, interpolation = cv2.INTER_AREA)
      	#results = pose.process(imgRGB)
      	#print(results.pose_landmarks)
      	#if results.pose_landmarks:
      		#mpDraw.draw_landmarks(img, results.pose_landmarks, mpPose.POSE_CONNECTIONS)
      	#for id, lm in enumerate(results.pose_landmarks.landmark):
      		#h, w,c = img.shape
      		#print(id, lm)
      		#cx, cy = int(lm.x*w), int(lm.y*h)
      		#cv2.circle(img, (cx, cy), 5, (255,0,0), cv2.FILLED)
      	#cTime = time.time()
      	#fps = 1/(cTime-pTime)
      	#pTime = cTime
      	#cv2.putText(img, str(int(fps)), (50,50), cv2.FONT_HERSHEY_SIMPLEX,1,(255,0,0), 3)
      	if cv2.getWindowProperty(window_title, cv2.WND_PROP_AUTOSIZE) >= 0:
      		cv2.imshow(window_title, frame)
      	else:
      		break
      	keyCode = cv2.waitKey(10) & 0xFF
      	if keyCode == 27 or keyCode == ord('q'):
      		break
    finally:
    	video_capture.release()
    	cv2.destroyAllWindows()


if __name__ == "__main__":
    show_camera()
